src/contracts/core/Slasher.sol

```diff
@@ -1,102 +1,368 @@
+        });
+        return pendingScalingFactor;
+    }
 
-    function getCorrectValueForInsertAfter(address, uint32) external view returns (uint256) {}
+    // TODO: this is a "naive" search since it brute-force backwards searches; we might technically want a binary search for completeness
+    /**
+     * @notice gets the scaling factor for the given operator and strategy at the given epoch
+     * @param operator the operator to get the scaling factor for
+     * @param strategy the strategy to get the scaling factor for
+     * @param epoch the epoch to get the scaling factor for
+     * @return the scaling factor for the given operator and strategy at the given epoch
+     */
+    function shareScalingFactorAtEpoch(
+        address operator,
+        IStrategy strategy,
+        uint32 epoch
+    ) public view returns (uint64) {
+        uint64 scalingFactor = SlashingAccountingUtils.SHARE_CONVERSION_SCALE;
+        (uint32 lookupEpoch, bool found) = _getLookupEpoch(operator, strategy, epoch);
+        if (found) {
+            scalingFactor = slashingRequests[operator][strategy][lookupEpoch].scalingFactor;
+        }
+        return scalingFactor;
+    }
 
-    function canWithdraw(
-        address,
-        uint32,
-        uint256
-    ) external returns (bool) {}
+    // @notice Returns the epoch in which the operator was last slashed
+    function lastSlashed(address operator, IStrategy strategy) public view returns (uint32) {
+        uint256 slashedEpochHistoryLength = slashedEpochHistory[operator][strategy].length;
+        if (slashedEpochHistoryLength == 0) {
+            // TODO: consider if a different return value is more appropriate for this special case
+            return 0;
+        } else {
+            return slashedEpochHistory[operator][strategy][slashedEpochHistoryLength - 1];
+        }
+    }
 
-    function operatorToMiddlewareTimes(
-        address,
-        uint256
-    ) external view returns (MiddlewareTimes memory) {}
+    function _modifyRequestedBipsToSlash(
+        address operator,
+        bytes4 operatorSetID,
+        IStrategy[] memory strategies,
+        uint32 epoch,
+        int32 bipsToModify
+    ) internal {
+        require(bipsToModify != 0, "Slasher._modifyRequestedBipsToSlash: cannot modify slashing by 0");
 
-    function middlewareTimesLength(address) external view returns (uint256) {}
+        IOperatorSetManager.OperatorSet memory operatorSet =
+            IOperatorSetManager.OperatorSet({avs: msg.sender, id: operatorSetID});
+        bytes32 operatorSetHash = _hashOperatorSet(operatorSet);
+        for (uint256 i = 0; i < strategies.length; ++i) {
+            IStrategy strategy = strategies[i];
+            uint32 requestedSlashedBipsBefore = requestedSlashedBips[operator][strategy][epoch][operatorSetHash];
+            int32 requestedSlashedBipsAfter = int32(requestedSlashedBipsBefore) + bipsToModify;
```

Comment:
BUG: this is *not* dealing correctly with the case of `requestedSlashedBipsAfter` exceeding 10,000!
see similar comment on the closed PR here https://github.com/Layr-Labs/eigenlayer-contracts/pull/543#discussion_r1651198852